reactive/kubernetes_worker.py

```diff
@@ -1325,3 +1331,19 @@ def get_registry_location():
         registry = ""
 
     return registry
+
+
+@when('coredns.ready', 'kube-control.dns.available')
+def set_cluster_dns():
+    """Set ClusterDNS in the kubelet config.
+
+    If a relation with the CoreDNS k8s charm is added its service
+    IP will be added in the kubelet config as the ClusterDNS argument
+    and the kubelet will be restarted.
+    """
+    kube_control = endpoint_from_flag('kube-control.dns.available')
+    dns = kube_control.get_dns()
+    ingress_ip = get_ingress_address(kube_control.endpoint_name)
+    configure_kubelet(dns, ingress_ip)
+    restart_unit_services()
+    update_kubelet_status()
```

Comment:
I recommend not calling `configure_kubelet`, `restart_unit_services`, and `update_kubelet_status` directly, because they make a lot of assumptions about how and when they are called. For example, `configure_kubelet` assumes [here](https://github.com/charmed-kubernetes/charm-kubernetes-worker/blob/1ab8da30048f4b06e75344d88d648b9a0c91412e/reactive/kubernetes_worker.py#L640) that the `endpoint.container-runtime.available` flag is set, but your `set_cluster_dns` handler does not guarantee that. That could result in a hook failure.

Use `set_state('kubernetes-worker.restart-needed')` instead.